### PR TITLE
Dm 2300

### PR DESCRIPTION
**Pull request by JoshingYou1**
### JIRA issue link
https://agile6.atlassian.net/browse/DM-2300

## Description - what does this code do?
Adds monthly metrics data to `metrics` downloadable excel spreadsheet.

## Testing done - how did you test it/steps on how can another person can test it 
1. Open two new windows in your browser.
2. Log in as an admin.
3. In one window, browse to the metrics tab in the dashboard.
4. Click on the `Metrics` tab and then click on `Export as .xlsx` at the top-right of the page.

**_Monthly Site Visits_**
1. Open the spreadsheet file and under the `General Traffic` header, there should be a sub-header called `Site Visits per Month`.
2. Look at the visit count for the month of `January 2021`.
3. Go to the other window and visit a page within DM(excluding the admin panel).
4. Now refresh the window with the admin panel, download a new spreadsheet, and check to make sure the visit count increased accordingly.

_**Monthly Practice Views**_
1. In the same spreadsheet as above, scroll down a bit to the `Practice Engagement and Commitment` header.
2. You should now see a new sub-header called `Practice Views per Month`.
3. Scroll to the right to get to the month of `January 2021`.
4. Pick any practice and look at the visit count.
5. In the other window, visit the show page of that same practice.
6. Refresh the admin panel page, go the the `Metrics` tab, and download the spreadsheet again.
7. Confirm that the visit count increased accordingly.

_**Monthly User Statistics**_
1. Within the same spreadsheet, scroll down to the very bottom and make sure there's a new section with a header named `User Statistics`.
2. Scroll to the right so you're looking at the month of `January 2021`.
3. First look at the `New users` row and remember the count.
4. In your database, create a new user, and make sure to include a date for the `confirmed_at` column and add `true` for the `skip_va_validation` column.
5. Now refresh the admin panel page, download the spreadsheet again, and check to make sure the count increased accordingly.
6. In that same spreadsheet, find the `Total users` row, which is located just below the `New users` row.
7. scroll to the right and remember the count for the month of `January 2021`.
8. Go to the other window and log out of your current profile.
9. Now, using one of the accounts from the `seed.rb` file, log into DM.
10. Go back to the admin panel window, refresh the page, download the spreadsheet again, and confirm that the `Total users` count has increased accordingly.

_**BONUS**_
1. Within the `General Traffic` header of both the metrics tab of the admin panel and the spreadsheet, make sure that `Unique Visitors` and `Number of Site Visits` have `(last month)` succeeding them.
2. Make sure `Total Accounts` has `(all-time)` succeeding it.

## Screenshots, Gifs, Videos from application (if applicable)
Screen Shot 2021-01-25 at 1 01 11 PM
Screen Shot 2021-01-25 at 1 01 39 PM
Screen Shot 2021-01-25 at 1 01 51 PM
Screen Shot 2021-01-25 at 12 59 56 PM

## Acceptance criteria
- [Views by Practice Page (Broken out by month--all months, maybe by pull into Excel, when you pull it into excel you get all the months)]

- [Site Visits (Broken out by month) to excel]

- [Total users per month to excel]

- [New users by month to excel]

- [Question:

            Going to go back to the month 2020, so whatever year we’re on it goes back one year

            Check on first of the month vs. day

            Want to start on first of the. month historically (e.g. start Jan 1 2020 through Jan. 19 2021)

            Have past 12 months, how is this going to work, do we archive January of 2020?

            12 months to the day or to the first of month?]

## Definition of done
- [ ] Unit tests written (if applicable)
- [ ] e2e/accessibility tests written (if applicable)
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [X] A link has been provided to the originating JIRA issue
- [X] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

File: app/admin/dashboard.rb [hunk 1]
```diff
@@ -47,6 +47,104 @@ def set_dashboard_values
...
+          practice_visits_by_month = []
+          pr_visit_ct = Ahoy::Event.where_props(practice_id: p[:id]).where(time: beg_of_month...end_of_month).count
+          practice_visits_by_month << pr_visit_ct
+          practice_visits_by_month.each do |visit_count|
+            @practice_views_array << [p.id, visit_count]
+          end
+        end
+
+        # Get site visits by month
+        site_visit_by_ip = Ahoy::Event.where(name: 'Site visit').where("properties->>'ip_address' is not null").where(time: beg_of_month..end_of_month).group("properties->>'ip_address'").count
```
Comment (emyl3) [inline app/admin/dashboard.rb:76]: I think we are reusing this somewhere else in the file. Can we make this into a method that we can call?

File: app/admin/dashboard.rb [hunk 2]
```diff
@@ -47,6 +47,104 @@ def set_dashboard_values
...
+
+      # Add user stats hash to make it easier to format spreadsheet data
+      @user_statistics = {
+          new_users: @new_users_by_month,
+          total_users: @total_users_by_month
+      }
+    end
+
+    def add_header_row_for_month_and_year(sheet, first_column_text, array, row_style)
+      sheet.add_row [
```
Comment (emyl3) [inline app/admin/dashboard.rb:111]: could we do something like
```
sheet_row = ["#{first_column_text}"] + array.map { |a| a.join(' ')}
sheet.add_row sheet_row, style: row_style
```

File: app/admin/dashboard.rb [hunk 3]
```diff
@@ -63,26 +161,49 @@ def export_metrics
...
           sheet.add_row [""], style: xlsx_divider
 
           sheet.add_row ["Practices"], style: xlsx_sub_header_1
           @practices_added_stats.each { |key, value| sheet.add_row [key.to_s.tr!('_', ' ').titleize, value], style: xlsx_entry }
           sheet.add_row [""], style: xlsx_divider
 
           sheet.add_row ["Practice Engagement & Commitment"], style: xlsx_sub_header_1
+          sheet.add_row ['Practice Views per Month'], style: xlsx_sub_header_2
+          add_header_row_for_month_and_year(sheet, 'Practice name', @month_and_year_array, xlsx_sub_header_3)
+          @practice_views_by_month.in_groups_of(13).each do |practice_views|
```
Comment (emyl3) [inline app/admin/dashboard.rb:187]: i think you can do something like
```
@practice_views_by_month.in_groups_of(13) do |practice_views|
    sheet_row = ["#{practice_views[0][0]}"] + practice_views.map { |pv| pv[1]}
    sheet.add_row sheet_row, style: xlsx_entry
end
```

File: app/views/home/index.html.erb [hunk 4]
```diff
@@ -3,6 +3,7 @@
 <% end %>
 
 <div class="dm-homepage">
+  <%= @test11 %>
```
Comment (emyl3) [inline app/views/home/index.html.erb:6]: do we need this?

File: app/admin/dashboard.rb [hunk 5]
```diff
@@ -47,6 +47,104 @@ def set_dashboard_values
...
+          array[10].join(' '),
+          array[11].join(' '),
+          array[12].join(' ')
+      ], style: row_style
+    end
+
+    def add_monthly_array_values_to_columns(hash, sheet, row_style)
+      hash.each do |key, value|
+        sheet.add_row [
+            key.to_s.tr!('_', ' ').titleize,
```
Comment (emyl3) [inline app/admin/dashboard.rb:132]: i think we can clean this up similarly to the above comment

**Review by emyl3 — COMMENTED** [review]
Let me know if my suggestions work for you haha 😅 Great work on this, Josh!

File: app/admin/dashboard.rb (hunk 1 above)
Comment (JoshingYou1) [reply app/admin/dashboard.rb:76]: I created a method for this. Thank you!

File: app/admin/dashboard.rb (hunk 2 above)
Comment (JoshingYou1) [reply app/admin/dashboard.rb:111]: Yes!! I adjusted the code. Thanks!

File: app/views/home/index.html.erb (hunk 4 above)
Comment (JoshingYou1) [reply app/views/home/index.html.erb:6]: Haha no. I removed it :)

File: app/admin/dashboard.rb (hunk 5 above)
Comment (JoshingYou1) [reply app/admin/dashboard.rb:132]: Sure can! Code has been updated! Thank you!

**Review by emyl3 — APPROVED** [review]

**Pull request merged**